File: lib/to_mdx/detectors/code_detector.rb
# frozen_string_literal: true

require_relative "./language_detector"
require_relative "../parsers/open_document/base"

module ToMdx
  module Detectors
    module CodeDetector
      # If this element looks like a code snippet, return the MDX for it
      def maybe_code_snippet_mdx
        return nil unless @context[:options][:use_heuristics]
        return nil unless is_using_monospace_font?

        text_content = to_s.chomp
        language = ToMdx::Detectors::LanguageDetector.detect(text_content)

        "\n```#{language}\n#{text_content}\n```\n"
      end

      def is_using_monospace_font?
        return false unless @element.respond_to?(:children)
        return false unless @context[:styles]

        @element.children.any? do |child|
          if child.attributes["text:style-name"]
            style_properties = @context[:styles].properties(child.attributes["text:style-name"])
            font_name = style_properties["font-family"]
            next true if font_name&.include?("monospace")
          end

          child.children.any? do |grandchild|
            if grandchild.attributes["text:style-name"]
              style_properties = @context[:styles].properties(grandchild.attributes["text:style-name"])
              font_name = style_properties["font-family"]
              font_name&.include?("monospace")
            end
          end
        end
      end
    end
  end
end

File: src/detectors/code-detector.ts
import type { ParsedElement } from "../parsers/base_element.ts";
import { LanguageDetector } from "./language-detector.ts";

export class CodeDetector {
  private element: ParsedElement;
  private toString: () => string;

  constructor(element: ParsedElement, toString: () => string) {
    this.element = element;
    this.toString = toString;
  }

  isCodeCandidate(): boolean {
    if (!this.element.context("options")?.useHeuristics) {
      return false;
    }

    return this.isUsingMonospaceFont();
  }

  maybeCodeSnippet(): string | null {
    if (!this.isCodeCandidate()) {
      return null;
    }

    const textContent = this.toString().trim();
    const language = this.detectLanguage(textContent);
    if (!language) {
      return null;
    }

    return `\n\`\`\`${language}\n${textContent}\n\`\`\`\n`;
  }

  private detectLanguage(textContent: string): string | null {
    const regexLanguage = LanguageDetector.detect(textContent);
    if (regexLanguage) {
      return regexLanguage;
    }

    if (!this.element.context("options")?.useLlmDetection) {
      return null;
    }

    return this.element.context("codeLanguageCache")?.get(textContent) ?? null;
  }

  isUsingMonospaceFont(): boolean {
    for (const child of this.element) {
      const styleName = child.attr("text:style-name");
      if (styleName) {
        const styleProperties = this.element.context("styles").get(styleName);
        const fontName = styleProperties["font-family"];
        if (fontName?.includes("monospace")) {
          return true;
        }
      }

      for (const grandchild of child) {
        const grandStyleName = grandchild.attr("text:style-name");
        if (grandStyleName) {
          const grandStyleProperties = this.element.context("styles").get(grandStyleName);
          const grandFontName = grandStyleProperties["font-family"];
          if (grandFontName?.includes("monospace")) {
            return true;
          }
        }
      }
    }

    return false;
  }
}
